src/libcore/num/f32.rs

```diff
@@ -257,7 +257,14 @@ impl Float for f32 {
 
     #[inline]
     fn powi(self, n: i32) -> f32 {
-        unsafe { intrinsics::powif32(self, n) }
+        let (mut r, mut a) = (1., self);
+        let mut b = if n < 0 { n.wrapping_neg() as u32 } else { n as u32 };
+        loop {
```

Comment:
I would personally prefer this as the following, which is simpler (IMO):

```rust
while b != 0 {
     if b % 2 != 0 { r *= a }
     b /= 2;
     a *= a;
}
return if n < 0 { r.recip() } else { r }
```

But I don't feel particularly strongly. (Yes, I know the code as written does an additional `a *= a`.)

If you feel strongly about avoiding that extra `a *= a` I even prefer this form (i.e. basically just avoid folding the `return` into a complicated expression in the loop):

```rust
loop {
     if b % 2 != 0 { r *= a }
     b /= 2;
     if b == 0 { break }
     a *= a;
}
return if n < 0 { r.recip() } else { r }
```